src/FsAutoComplete/LspServers/FSharpLspClient.fs

```diff
@@ -143,74 +144,164 @@ type ServerProgressReport(lspClient: FSharpLspClient, ?token: ProgressToken) =
-              let fileName = IO.Path.GetFileName message
-
-              if inflightEvents.TryAdd(eventArgs.Task, progressReport) then
-                progressReport.Begin($"Dependent Typecheck {fileName}", message = message)
-                |> Async.Start
-            | 6 ->
-              let message = eventArgs.Payload.[0] |> string
-
-              match inflightEvents.TryRemove(eventArgs.Task) with
-              | true, report ->
-                report.End($"Finished {message}") |> Async.Start
-                tryDispose report
-                ()
-              | false, _ ->
-
-                ()
-            | other -> ()
-
-          message
-      with e ->
-        ()
-
-  member _.DisableEvents(source) = ``base``.DisableEvents(source)
-
-  interface System.IDisposable with
-    member this.Dispose() =
-      lock this
-      <| fun () ->
-        if isNull source then () else this.DisableEvents(source)
-        isDisposed <- true
-        inflightEvents.Values |> Seq.iter (tryDispose)
-        inflightEvents <- null
+  let inflightEvents =
+    ConcurrentDictionary<_, System.Diagnostics.Activity * ServerProgressReport>()
+
+  let isStopped (activity: Activity) =
+#if NET6_0
+    false
+#else
+    activity.IsStopped
+#endif
+    // giving this 1 seconds to report something, otherwise assume it's a dead activity
+    || ((DateTime.UtcNow - activity.StartTimeUtc) > TimeSpan.FromSeconds(1.)
+        && activity.Duration = TimeSpan.Zero)
+
+  let getTagItemSafe key (a: Activity) = a.GetTagItem key |> Option.ofObj
+
+  let getFileName =
+    getTagItemSafe Tracing.SemanticConventions.FCS.fileName
+    >> Option.map string
+    >> Option.map IO.Path.GetFileName
+    >> Option.defaultValue String.Empty
+
+  let getProject =
+    getTagItemSafe Tracing.SemanticConventions.FCS.project
+    >> Option.map string
+    >> Option.map IO.Path.GetFileName
+    >> Option.defaultValue String.Empty
+
+
+  let getUserOpName =
+    getTagItemSafe Tracing.SemanticConventions.FCS.userOpName
+    >> Option.map string
+    >> Option.defaultValue String.Empty
+
+  let mbp =
+    MailboxProcessor.Start(fun inbox ->
+      async {
+        while not isDisposed do
+          for (a, p) in inflightEvents.Values do
+            // We don't get always get a corresponding ActivityStopped for a given Activity event so this looks at anything that hasn't had a duration incresed at all
+            // We don't seem to get a corresponding ActivityStopped on cancelled typechecks.
+            if isStopped a then
+              do! p.End()
+              inflightEvents.TryRemove(a.Id) |> ignore
+            else
+              // FSC doesn't start their spans with tags so we have to see if it's been added later https://github.com/dotnet/fsharp/issues/14776
+              let fileName = getFileName a
+              let userOpName = getUserOpName a
+              do! p.Report(message = $"{fileName} - {userOpName}")
```

Comment:
As noted, FCS doesn't start their spans with tags, so we can't get them immediately to report on.  This will update our notifications with tag info if it's still around.